test/compilable/issue21203.d

```diff
@@ -0,0 +1,144 @@
+    }
+}
+
+extern(C++, "ns")
+{
+    class MyClassB {}
+    void funf(ScopeClass!MyClassB);           // mangles MyClass
+    void fung(const ScopeClass!MyClassB);     // mangles const MyClass
+    void funh(ref ScopeClass!MyClassB);       // mangles MyClass&
+    void funi(ref const ScopeClass!MyClassB); // mangles const MyClass&
+    void funj(const(ScopeClass!MyClassB)*);
+}
+
+version (Posix)
+{
+    static assert(funf.mangleof == "_ZN2ns4funfENS_8MyClassBE");
+    static assert(fung.mangleof == "_ZN2ns4fungENS_8MyClassBE");
+    static assert(funh.mangleof == "_ZN2ns4funhERNS_8MyClassBE");
+    static assert(funi.mangleof == "_ZN2ns4funiERKNS_8MyClassBE");
+    static assert(funj.mangleof == "_ZN2ns4funjEPKNS_8MyClassBE");
+}
+else version (CppRuntime_Microsoft)
+{
+    static assert(funf.mangleof == "?funf@ns@@YAXVMyClassB@1@@Z");
+    static assert(fung.mangleof == "?fung@ns@@YAXVMyClassB@1@@Z");
+    static if (size_t.sizeof == ulong.sizeof)
+    {
+        static assert(funh.mangleof == "?funh@ns@@YAXAEAVMyClassB@1@@Z");
+        static assert(funi.mangleof == "?funi@ns@@YAXAEBVMyClassB@1@@Z");
+        static assert(funj.mangleof == "?funj@ns@@YAXPEBVMyClassB@1@@Z");
+    }
+    else
+    {
+        static assert(funh.mangleof == "?funh@ns@@YAXAAVMyClassB@1@@Z");
+        static assert(funi.mangleof == "?funi@ns@@YAXABVMyClassB@1@@Z");
+        static assert(funj.mangleof == "?funj@ns@@YAXPBVMyClassB@1@@Z");
+    }
+}
+
+extern(C++)
+{
+    void funTempl(T)();
+    class MyClassC {}
+    alias funTemplA = funTempl!(ScopeClass!MyClassC);
+    alias funTemplB = funTempl!(const ScopeClass!MyClassC);
+    alias funTemplC = funTempl!(const(ScopeClass!MyClassC)*);
+    // N.B funTempl!([const] ref ScopeClass!MyClassC) is not permissable in D
+}
+version (Posix)
+{
+    static assert(funTemplA.mangleof == "_Z8funTemplI8MyClassCEvv");
+    static assert(funTemplB.mangleof == "_Z8funTemplIK8MyClassCEvv");
+    static assert(funTemplC.mangleof == "_Z8funTemplIPK8MyClassCEvv");
+}
+else version (CppRuntime_Microsoft)
+{
+    static assert(funTemplA.mangleof == "??$funTempl@VMyClassC@@@@YAXXZ");
+    static assert(funTemplB.mangleof == "??$funTempl@$$CBVMyClassC@@@@YAXXZ");
+    static if (size_t.sizeof == ulong.sizeof)
+        static assert(funTemplC.mangleof == "??$funTempl@PEBVMyClassC@@@@YAXXZ");
+    else
+        static assert(funTemplC.mangleof == "??$funTempl@PBVMyClassC@@@@YAXXZ");
+}
+
+extern(C++, "std")
+{
+    extern(C++, struct)
+    pragma(mangle, "function")
+    class _function(F)
+    {
+
+    }
+}
+
+template FunctionOf(F)
+{
+    F f;
+    alias FunctionOf = typeof(*f);
+}
+extern(C++) void funk( ScopeClass!(_function!(FunctionOf!(void function(int))),"function") a ){ /*a.ptr(1);*/ }
+
+version (Posix)
+{
+    static assert(funk.mangleof == "_Z4funkSt8functionIFviEE");
+}
+else version (CppRuntime_Microsoft)
+{
+    static assert(funk.mangleof == "?funk@@YAXV?$function@$$A6AXH@Z@std@@@Z");
+}
+
+
```

Comment:
Can we clean the trailing (or leading) whitespace ? Some people have configured their editor to automatically remove it so it's never an issue :)